File: slc.php
<?php
/*
Copyright (C) 2026 Andrew Lavens

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://gnu.org>.
*/


function slc(Array $arr, Int $max_cons, Int|Null $offset = NULL, Int $length = 1): Array
{
    $passed = false;
    while (!$passed) {
        $collection = $arr;
        $n_values = count($collection);
        $result = [];
        $current_cons = 1;
        foreach(range(1, $n_values) as $idx) {
            $suitable_choice = false;
            while (!$suitable_choice) {
                $choice_idx = array_rand($collection);
                $choice = $collection[$choice_idx];
                if (count($result) > 0) {
                    if ($offset) {
                        $current = substr($choice, $offset, $length);
                        $previous = substr($result[count($result)-1], $offset, $length);
                    } else {
                        $current = $choice;
                        $previous = $result[count($result)-1];
                    }
                    if ($current_cons == $max_cons && count($result) > 0 && $current == $previous) {
                        $suitable_choice = false;
                        if ($offset) {
                            $substrings = [];
                            foreach($collection as $item) {
                                $substrings[] = substr($item, $offset, $length);
                            }
                            $keys = array_unique($substrings);
                            if (count($keys) == 1 && $keys[0] == substr($result[count($result)-1], $offset, $length)) {
                                break;
                            }
                        } else {
                            $keys = array_unique($collection);
                            if (count($keys) == 1 && $keys[0] == $result[count($result)-1]) {
                                break;
                            }
                        }
                    } else {
                        $suitable_choice = true;
                    }
                } else {
                    $suitable_choice = true;
                }
            }
            $result[] = $choice;
            array_splice($collection, $choice_idx, 1);
            if (count($result) > 1 && $current == $previous) {
                $current_cons++;
            } else {
                $current_cons = 1;
            }
        }
        if (count($collection) == 0) {
            $passed = true;
        }
    }
    return $result;
}

$arr = [1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4];
print_r(implode(',', slc($arr, 1))."\n\n");
print_r(implode(',', slc($arr, 2))."\n\n");
print_r(implode(',', slc($arr, 3))."\n\n");
print_r(implode(',', slc($arr, 4))."\n\n");
print_r(implode(',', slc($arr, 5))."\n\n");

File: slc.py
'''
Copyright (C) 2026 Andrew Lavens

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://gnu.org>.
'''
import random

def slc(arr, max_cons, offset = None, length = 1):
    passed = False
    while not passed:
        collection = arr.copy()
        substrings = []
        if offset:
            for item in collection:
                substrings.append(item[offset:offset+length])
        n_values = len(collection)
        result = []
        current_cons = 1
        for idx in range(n_values):
            suitable_choice = False
            while not suitable_choice:
                choice = random.choice(collection)
                if len(result) > 0:
                    if offset:
                        current = choice[offset:offset+length]
                        previous = result[-1][offset:offset+length]
                    else:
                        current = choice
                        previous = result[-1]
                    if current_cons == max_cons and current == previous:
                        suitable_choice = False
                        if offset:
                            keys = list(dict.fromkeys(substrings))
                            if len(keys) == 1 and keys[0] == result[-1][offset:offset+length]:
                                break
                        else:
                            keys = list(dict.fromkeys(collection))
                            if len(keys) == 1 and keys[0] == result[-1]:
                                break
                    else:
                        suitable_choice = True
                else:
                    suitable_choice = True
            result.append(choice)
            collection.pop(collection.index(choice))
            if len(substrings) > 0:
                substrings.pop(substrings.index(choice[offset:offset+length]))
            if len(result) > 1:
                if current == previous:
                    current_cons += 1
                else:
                    current_cons = 1
        if len(collection) == 0:
            passed = True
    return result        


'''
arr = [1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3,3,3,4,4,4,4,4,4,4,4,4,4]
print ("Array: ", arr)
print ("One", slc(arr, 1))
print ("Two", slc(arr, 2))
print ("Three", slc(arr, 3))
print ("Four", slc(arr, 4))
print ("Five", slc(arr, 5))
'''
